### PR TITLE
Fuzz Network Test Framework

### LINKED ISSUE #320: Create a testing framework that generates random test cases
We need a test framework that creates random tests and simulates scenarios where nodes either crash or go offline and recover, while transactions are being sent to the nodes.

For that, we need a simple and trivial memory pool that we can inject fake transactions into. 
Injecting a transaction into the memory pool will be done by adding a transaction to the memory pool of each instance.
Once a block is indexed in storage, the memory pool is cleaned from the transactions in the block. 

For a max round window $m$, we can define a random length for a test - $l=m\cdot k$ for some $k>1$. This will be the length of the number of blocks that need to be finalized. 

The test will have a random number of nodes - $3<n<10$ and will do two things in parallel:

1. Inject transactions in a slow and controlled manner until $l$ blocks are finalized.
2. Divide the time to intervals of $T$ seconds (1-2 seconds), and will randomly crash or disconnect up to $f$ nodes in an interval. 

The first stage of each test will be generating a scenario **descriptor**, in some human readable format (e.g a JSON array) and then feeding it into a **driver** module that will orchestrate the nodes and network according to the descriptor. 

The test will log what it is doing at each time interval, as well as the status of all nodes.

The motivation is that by creating random test scenarios, we might uncover edge cases and will then later be able to re-run them by taking the descriptor and feeding it into the test driver module.

### PR DESCRIPTION
**Pull request by samliok**

### CODE REVIEW

File: testutil/random_network/mempool.go [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+	}
+
+	if err := tx.Verify(ctx); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (m *Mempool) isTxInChain(txID txID, parentDigest simplex.Digest) bool {
+	block, exists := m.verifiedButNotAcceptedTXs[parentDigest]
```
Comment (yacovm) [inline testutil/random_network/mempool.go]: I'm not sure this works for a case where we have a leader failover that creates a sibling block.

Suppose block `a` is notarized and then a block `b` is built on top of it and broadcast, and verified by a node but then the majority notarize the empty block.

In the next round, block `b'` is proposed on top of block `a`. 
We need a way to rollback the verified but not accepted map for the parent `a` because both blocks `b` and `b'` are built on top of it.

One way of doing it, is move this book-keeping to the blocks themselves. 
Specifically, have the instances of the blocks maintain pointers to blocks they are built upon, and then just have the transaction list inside a block a map, where when we encode the bytes of the block, we do it across some order that is set at the time we build the block.

**Comment by yacovm** [comment]
Can we rebase on top of main?

File: testutil/random_network/network.go [hunk 2]
```diff
@@ -0,0 +1,361 @@
...
+		randomness:           r,
+		l:                    l,
+		t:                    t,
+		lock:                 sync.Mutex{},
+		config:               config,
+	}
+}
+
+func (n *Network) StartInstances() {
+	panic("Call Run() Instead")
```
[the remark this answers is not in the document]
Comment (samliok) [reply testutil/random_network/network.go]: because `BasicNetwork` is embedded in `Network` i don't want the caller accidentally calling `StartInstances` thinking its a valid method

File: testutil/random_network/mempool.go (hunk 1 above)
Comment (samliok) [reply testutil/random_network/mempool.go]: my hope for the mempool implementation was for it be simple but useful enough. so we can ensure a tx we issued got accepted, but things like more complicated tree rollbacks were out of scope for the test

File: testutil/random_network/network.go [hunk 3]
```diff
@@ -0,0 +1,325 @@
...
+		if height < minHeight {
+			minHeight = height
+			minHeightNodeID = node.E.ID
+		}
+	}
+
+	return minHeightNodeID
+}
+
+func (n *Network) recoverToHeight(height uint64) {
```
Comment (yacovm) [inline testutil/random_network/network.go]: The below way we are more deterministic (written by me, not thanks to CC):  

```
func (n *Network) recoverToHeight(height uint64) {
	// Decide which crashed nodes to recover and in what order.
	// This keeps all randomness calls deterministic (fixed count per invocation),
	// independent of the non-deterministic AdvanceTime loop below.
	var nodesToRecover []int
	for i, node := range n.nodes {
		if node.isCrashed.Load() {
			nodesToRecover = append(nodesToRecover, i)
		}
	}
	n.randomness.Shuffle(len(nodesToRecover), func(i, j int) {
		nodesToRecover[i], nodesToRecover[j] = nodesToRecover[j], nodesToRecover[i]
	})

	// Recover nodes one at a time in the chosen order, advancing time until
	// all nodes have caught up before recovering the next one.
	for _, idx := range nodesToRecover {
		n.logger.Debug("Recovering node", zap.Stringer("nodeID", n.nodes[idx].E.ID))
		n.startNode(idx)

		for n.nodes[idx].storage.NumBlocks() < height {
			n.logger.Debug("Advancing network time", zap.Uint64("num crashed nodes", n.numCrashedNodes()),
				zap.Uint64("min height", n.getMinHeight()),
				zap.Uint64("max height", n.getMaxHeight()),
				zap.Stringer("Smallest node ID", n.getMinHeightNodeID()),
				zap.Uint64("target height", height),
			)

			for _, node := range n.nodes {
				n.lock.Lock()
				node.AdvanceTime(n.config.AdvanceTimeTickAmount)
				n.lock.Unlock()
			}
	
		}
	}

	// Advance time until all remaining nodes reach the target height.
	for n.getMinHeight() < height {
		n.logger.Debug("Advancing network time", zap.Uint64("num crashed nodes", n.numCrashedNodes()),
			zap.Uint64("min height", n.getMinHeight()),
			zap.Uint64("max height", n.getMaxHeight()),
			zap.Stringer("Smallest node ID", n.getMinHeightNodeID()),
			zap.Uint64("target height", height),
		)

		n.lock.Lock()
		n.BasicInMemoryNetwork.AdvanceTime(n.config.AdvanceTimeTickAmount)
		n.lock.Unlock()

	}
}
```

**Comment by yacovm** [comment]
will continue the review tomorrow

File: testutil/random_network/mempool.go [hunk 4]
```diff
@@ -0,0 +1,324 @@
...
+	// Ensure the block has not already been verified or accepted
+	if _, exists := m.verifiedButNotAcceptedBlocks[b.digest]; exists {
+		return fmt.Errorf("%w: %s", errDoubleBlockVerification, b.digest)
+	}
+
+	if _, exists := m.acceptedBlocks[b.digest]; exists {
+		return fmt.Errorf("%w: %s", errDoubleBlockVerification, b.digest)
+	}
+
+	// Ensure the parent block is accepted or verified
```
Comment (yacovm) [inline testutil/random_network/mempool.go]: ditto, shouldn't we fail the entire test?

File: testutil/random_network/node.go [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+// this is important because blocks are not serialized/deserialized in our current comm implementation, so sending blocks
+// also sends relevant state associated with the node that is sending the message which can cause unintended side effects.
+func (n *Node) copyMessage(msg *simplex.Message) simplex.Message {
+	// Create a copy of the message to avoid mutating shared state in the in-memory network
+	msgCopy := *msg
+
+	switch {
+	case msgCopy.BlockMessage != nil:
+		block := msgCopy.BlockMessage.Block.(*Block)
+
```
Comment (yacovm) [inline testutil/random_network/node.go]: can we just have a function:
```
func (n *Node) copyBlock(b simplex.Block) simplex.Block {
	blockCopy := *b.(*Block)
	blockCopy.mempool = n.mempool
	return &blockCopy
}
```

and just call it everywhere?

File: testutil/random_network/network.go [hunk 6]
```diff
@@ -0,0 +1,325 @@
...
+		allAccepted := true
+		for _, node := range n.nodes {
+			if node.isCrashed.Load() {
+				continue
+			}
+			if accepted := node.areTxsAccepted(txs); !accepted {
+				node.mempool.lock.Lock()
+				n.logger.Debug("Not all txs accepted yet by node", zap.Stringer("nodeID", node.E.ID), zap.Int("unaccepted txs in mempool", len(node.mempool.unacceptedTxs)))
+				node.mempool.lock.Unlock()
+				allAccepted = false
```
[the remark this answers is not in the document]
Comment (samliok) [reply testutil/random_network/network.go:180]: why can we break here? this is when allAccepted is false and we want to break when all are accepted

File: testutil/random_network/config.go [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+	"github.com/ava-labs/simplex"
+)
+
+type FuzzConfig struct {
+	// The minimum and maximum number of nodes in the network.
+	MinNodes int // Default is 3.
+	MaxNodes int // Default is 10.
+
+	// The minimum and maximum number of transactions to be issued at a block. Default is between 5 and 20.
+	MinTxsPerIssue int
```
[the remark this answers is not in the document]
Comment (samliok) [reply testutil/random_network/config.go:18]: i was hoping that we would occasionally issue more than a blocks worth of transactions at a time. this way we when we are waiting for txs to be accepted we need to wait multiple rounds. Potentially even more rounds than 2 if nodes are blacklisted. This way we hit more code paths and hopefully more edge cases.

File: testutil/random_network/mempool.go [hunk 8]
```diff
@@ -0,0 +1,324 @@
...
+		}
+	}
+}
+
+// VerifyBlock verifies the block and its transactions. Errors if any tx is invalid or if there are duplicate txs in the block.
+func (m *Mempool) VerifyBlock(ctx context.Context, b *Block) error {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	// Ensure the block has not already been verified or accepted
```
[the remark this answers is not in the document]
Comment (samliok) [reply testutil/random_network/mempool.go:107]: added error logs which should fail the test

File: testutil/random_network/node.go [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+	return n
+}
+
+func (n *Node) HandleMessage(msg *simplex.Message, from simplex.NodeID) error {
+	msgCopy := n.copyMessage(msg)
+	return n.BasicNode.HandleMessage(&msgCopy, from)
+}
+
+// copyMessage creates a copy of the message and its relevant fields to avoid mutating shared state in the in-memory network
```
[the remark this answers is not in the document]
Comment (samliok) [reply testutil/random_network/node.go:103]: yep

File: testutil/random_network/tx.go [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+	ShouldFailVerification bool
+}
+
+func (aTX asn1TX) toTX() *TX {
+	var idArr txID
+	copy(idArr[:], aTX.ID)
+	return &TX{ID: idArr, shouldFailVerification: aTX.ShouldFailVerification}
+}
+
+func CreateNewTX() *TX {
```
[the remark this answers is not in the document]
Comment (samliok) [reply testutil/random_network/tx.go:38]: eh, i think it's fine to not include our PRG here.  I've found block hashes aren't that helpful to keep deterministic imo, at least when i'm debugging.

File: testutil/random_network/mempool.go [hunk 11]
```diff
@@ -0,0 +1,324 @@
...
+	}
+
+	if len(m.unacceptedTxs) > 0 {
+		m.logger.Debug("After accepting block, moved txs back to unaccepted due to sibling/uncle blocks being purged", zap.Int("num unaccepted txs", len(m.unacceptedTxs)))
+		m.NotifyTxsReady()
+	}
+}
+
+// go through any blocks that build off of this one and move their txs
+func (m *Mempool) purgeChildren(block *Block) {
```
[the remark this answers is not in the document]
Comment (samliok) [reply testutil/random_network/mempool.go]: yep I think this commit fixes that issue 
https://github.com/ava-labs/Simplex/pull/326/commits/b2f99b122180c9d192b2b3f0af1ed0cb55b05aef

File: testutil/random_network/network.go (hunk 3 above)
Comment (samliok) [reply testutil/random_network/network.go:117]: > for _, node := range n.nodes {
				n.lock.Lock()
				node.AdvanceTime(n.config.AdvanceTimeTickAmount)
				n.lock.Unlock()
			}

why not just advance the time using 

```
		n.BasicInMemoryNetwork.AdvanceTime(n.config.AdvanceTimeTickAmount)
```

also i think this change is good, can we add it as a follow up though?

Comment (yacovm) [reply testutil/random_network/network.go:117]: sure

**Review by yacovm — APPROVED** [review]
